samza-core/src/main/java/org/apache/samza/zk/ZkBarrierForVersionUpgrade.java

```diff
@@ -19,161 +19,162 @@
- * from the notification and the barrier becomes invalid.
+ * ZkBarrierForVersionUpgrade is an implementation of distributed barrier, which guarantees that the expected barrier
+ * size and barrier participants match before marking the barrier as complete.
+ * It also allows the caller to expire the barrier.
+ *
+ * This implementation is specifically tailored towards barrier support during jobmodel version upgrades. The participant
+ * responsible for the version upgrade starts the barrier by invoking {@link #create(String, List)}.
+ * Each participant in the list, then, joins the new barrier. When all listed participants {@link #join(String, String)}
+ * the barrier, the creator marks the barrier as done which signals the end of barrier.
+ * The creator of the barrier can invoke {@link #expire(String)} to indicate that the barrier has timed_out or
+ * is no longer valid.
+ *
+ * The caller can listen to events associated with the barrier by registering a {@link ZkBarrierListener}.
+ *
+ * Zk Tree Reference:
+ * /barrierRoot/
+ *  |
+ *  |- barrier_{version1}/
+ *  |   |- barrier_state/
+ *  |   |  ([DONE|TIMED_OUT])
+ *  |   |- barrier_participants/
+ *  |   |   |- {id1}
+ *  |   |   |- {id2}
+ *  |   |   |-  ...
  */
-public class ZkBarrierForVersionUpgrade implements BarrierForVersionUpgrade {
-  private final ZkUtils zkUtils;
-  private final static String BARRIER_DONE = "done";
-  private final static String BARRIER_TIMED_OUT = "TIMED_OUT";
+public class ZkBarrierForVersionUpgrade {
   private final static Logger LOG = LoggerFactory.getLogger(ZkBarrierForVersionUpgrade.class);
+  private final ZkUtils zkUtils;
+  private final BarrierKeyBuilder keyBuilder;
+  private final Optional<ZkBarrierListener> barrierListener;
 
-  private final ScheduleAfterDebounceTime debounceTimer;
+  public enum State {
+    TIMED_OUT, DONE
+  }
 
-  private final String barrierPrefix;
-  private String barrierDonePath;
-  private String barrierProcessors;
-  private static final String VERSION_UPGRADE_TIMEOUT_TIMER = "VersionUpgradeTimeout";
-  private final long barrierTimeoutMS;
 
-  public ZkBarrierForVersionUpgrade(String barrierId, ZkUtils zkUtils, ScheduleAfterDebounceTime debounceTimer, long barrierTimeoutMS) {
+  public ZkBarrierForVersionUpgrade(String barrierRoot, ZkUtils zkUtils, ZkBarrierListener barrierListener) {
     if (zkUtils == null) {
       throw new RuntimeException("Cannot operate ZkBarrierForVersionUpgrade without ZkUtils.");
     }
     this.zkUtils = zkUtils;
-    barrierPrefix = zkUtils.getKeyBuilder().getJobModelVersionBarrierPrefix(barrierId);
-    this.debounceTimer = debounceTimer;
-    this.barrierTimeoutMS = barrierTimeoutMS;
-  }
-
-  protected long getBarrierTimeOutMs() {
-    return barrierTimeoutMS;
+    this.keyBuilder = new BarrierKeyBuilder(barrierRoot);
+    this.barrierListener = Optional.ofNullable(barrierListener);
```

Comment:
> Please use different names - having same name, but different types might be misleading.

Are you referring to keyBuilder variable name ?

> (besides, I am not sure on value of having an Optional here, since it doesn't protect you from calling .get() on this object and throws a confusing NoSuchElementException instead of NullPointerException. So it only changes ifNotNull for ifPresent...but ..ok)

I agree there is a possibility of NoSuchElementException when using Optionals. But I am not following what you are suggesting I change? Can you elaborate your suggestion?